File: CP6.Entity/DomainModels/Order.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CP6.Entity.DomainModels;

/// <summary>
/// 受注ヘッダー（MSBBPA070 第1页 基本情報のヘッダー部）
/// </summary>
/// <remarks>
/// 仕様書 §11.1 T_Order
/// 業務 PK：Web受注NO（シーケンス自動採番）
/// 1 受注 → 多 OrderDetail → 多 OrderProcess / OrderMaterial
/// </remarks>
[Table("T_Order")]
public class Order : BaseBizEntity
{
    // ───── 業務主キー ─────
    /// <summary>Web受注NO（業務 PK、シーケンス自動採番）</summary>
    [Required, MaxLength(20)]
    public string WebOrderNo { get; set; } = string.Empty;

    // ───── 顧客・受注区分 ─────
    /// <summary>得意先コード</summary>
    [Required, MaxLength(20)]
    public string CustomerCd { get; set; } = string.Empty;

    /// <summary>受注区分（汎用マスタ：加工製品/シート/原紙/購買品/商品/版型/有償支給/経費品/輪切り）</summary>
    [Required, MaxLength(4)]
    public string OrderType { get; set; } = string.Empty;

    /// <summary>受注部門</summary>
    [MaxLength(10)] public string? OrderDepartment { get; set; }

    /// <summary>受注日</summary>
    public DateTime? OrderDate { get; set; }

    /// <summary>客先納期</summary>
    public DateTime? CustomerDeliveryDate { get; set; }

    /// <summary>数量（伝票合計）</summary>
    [Column(TypeName = "decimal(21,8)")]
    public decimal? Quantity { get; set; }

    /// <summary>注文書NO</summary>
    [MaxLength(30)] public string? OrderSheetNo { get; set; }

    /// <summary>先方担当</summary>
    [MaxLength(50)] public string? CustomerContact { get; set; }

    /// <summary>宛名</summary>
    [MaxLength(100)] public string? Addressee { get; set; }

    /// <summary>運送会社</summary>
    [MaxLength(20)] public string? Carrier { get; set; }

    /// <summary>出荷日時（YYYY/MM/DD HH:MM 文字列）</summary>
    [MaxLength(16)] public string? ShipDateTime { get; set; }

    /// <summary>出荷条件</summary>
    [MaxLength(20)] public string? ShipCondition { get; set; }

    /// <summary>売価区分：1=個別単価 / 2=セット単価</summary>
    [MaxLength(1)] public string? SalesPriceDiv { get; set; }

    // ───── mcframe7 連携 ─────
    /// <summary>受注NO（mc 連携キー；新規=NULL or "*"）</summary>
    [MaxLength(20)] public string? McOrderNo { get; set; }

    /// <summary>ステータス：0=未転送 / 9=mcframe7 転送済</summary>
    public int Status { get; set; } = 0;

    /// <summary>mcframe7 転送FLG（FALSE=未転送）</summary>
    public bool McTransferFlg { get; set; } = false;

    // ───── 出荷実績（WMS出荷確定 → ERP回写 / Phase4）─────
    /// <summary>出荷ステータス：0=未出荷 / 5=一部出荷 / 9=出荷済</summary>
    public int ShipStatus { get; set; } = 0;

    /// <summary>実出荷日時（WMS出荷確定で記録。ユーザ入力の ShipDateTime とは別管理）</summary>
    public DateTime? ActualShipDate { get; set; }

    // ───── 受注ライフサイクル（Phase 6 取消链）─────
    /// <summary>
    /// 受注ライフサイクル状態：Confirmed / InProduction / Shipped / Cancelled / PartiallyCancelled
    /// </summary>
    /// <remarks>
    /// 既存の <see cref="Status"/>（mc転送）とは別軸の独立フィールド。
    /// 取値常量は <see cref="OrderLifecycleStatus"/> 参照。
    /// </remarks>
    [MaxLength(20)]
    public string OrderStatus { get; set; } = OrderLifecycleStatus.Confirmed;

    /// <summary>取消時刻（UTC、CancelAsync 成功時に記録）</summary>
    public DateTime? CancelledAt { get; set; }

    /// <summary>取消理由（必須 — 監査用、OperLog にも別途記録）</summary>
    [MaxLength(200)]
    public string? CancelReason { get; set; }

    // ───── メモ ─────
    [MaxLength(100)] public string? Memo1 { get; set; }
    [MaxLength(100)] public string? Memo2 { get; set; }
    [MaxLength(100)] public string? Memo3 { get; set; }

    // ───── ナビゲーションプロパティ ─────
    public List<OrderDetail> Details { get; set; } = new();
}

/// <summary>
/// <see cref="Order.OrderStatus"/> 受注ライフサイクル状態 取値常量（Phase 6）
/// </summary>
public static class OrderLifecycleStatus
{
    /// <summary>確定済（受注作成直後の既定値）</summary>
    public const string Confirmed = "CONFIRMED";

    /// <summary>製造中（MES に展開済 / WO 着手後）</summary>
    public const string InProduction = "IN_PRODUCTION";

    /// <summary>出荷済（WMS 出荷確定後、Phase 4 ErpBridge 回写経由）</summary>
    public const string Shipped = "SHIPPED";

    /// <summary>取消完了（全関連 WO/Outbound も取消済）</summary>
    public const string Cancelled = "CANCELLED";

    /// <summary>一部取消（半路状態 — 一部の WO/Outbound は既に着手のため未取消）</summary>
    public const string PartiallyCancelled = "PARTIALLY_CANCELLED";
}

File: cp6.web/src/types/order.ts
// MSBBPA070 / 080 / 090 Web 受注 类型定义
// 对应后端 CP6.Entity/DTOs/OrderDto.cs

import type { ApiResult } from '@/types/estimateCalc'

/** PA070 操作种别 */
export enum OrderOperationType {
  New = 10,
  Edit = 20,
  Delete = 30,
  View = 40,
}

/** PA070 受注 1 个完整 DTO（ヘッダー + 多明細 + 多工程 + 多材料） */
export interface OrderDto {
  webOrderNo?: string

  // ───── ヘッダー ─────
  customerCd: string
  orderType: string
  orderDepartment?: string
  orderDate?: string
  customerDeliveryDate?: string
  quantity?: number
  orderSheetNo?: string
  customerContact?: string
  addressee?: string
  carrier?: string
  shipDateTime?: string
  shipCondition?: string
  salesPriceDiv?: string
  memo1?: string
  memo2?: string
  memo3?: string

  details: OrderDetailDto[]

  status: number
  mcTransferFlg: boolean
  mcOrderNo?: string
  rowVersion?: string
}

/** 受注明細 1 行（製品単位）— 約 130 項目 */
export interface OrderDetailDto {
  webOrderNo?: string
  webOrderDetailNo: number

  haibaiNo1?: string
  haibaiNo2?: string
  haibaiNo3?: string
  haibaiNo4?: string

  productCd: string
  itemCd?: string
  branch1?: string
  branch2?: string
  branch3?: string

  productCatBig?: string
  productCatMid?: string
  productCatSml?: string

  customerItemName1?: string
  customerItemName2?: string
  customerPartNo?: string
  cpItemName1?: string
  cpItemName2?: string
  janCode?: string

  qtyUnit?: string
  quantity?: number
  specialPriceFlg?: string
  unitPriceUnit?: string
  setUnitPrice?: number
  individualUnitPrice?: number
  amount?: number

  deliveryCd?: string
  deliveryName?: string
  customerDeliveryDate?: string
  logisticsGroup?: string
  haibaiKbn?: string

  consignedSalesFlg?: string
  salesReason?: string
  consignedSalesQty?: number

  fscOrderType?: string
  fscProductDiv?: string
  fscMaterialDiv?: string
  fscManagementNo?: string
  foodSafety?: string

  shipInspection?: string
  fixedShipment?: string
  deliveryReserve?: number
  salesSample?: number
  salesAvailable?: string

  // 構成情報（13 フィールド）
  sheetFlute?: string
  paperCdF?: string
  paperCdC?: string
  paperCdB?: string
  printCdF?: string
  printCdC?: string
  printCdB?: string
  embossCdF?: string
  embossCdC?: string
  embossCdB?: string
  makerCdF?: string
  makerCdC?: string
  makerCdB?: string

  // 寸法
  sheetPrint?: string
  bladeWidth?: number
  bladeFlow?: number
  gutterFb?: number
  gutterLr?: number
  sheetDimW?: number
  sheetDimF?: number
  salesWidth?: number
  finalMachineProcess?: string

  // 備考
  printNote?: string
  mfgNote?: string
  remfgNote?: string
  slipNote?: string
  deliveryNote?: string
  shipNote1?: string
  shipNote2?: string

  // 仕入情報
  defectiveHaibaiNo?: string
  purchaseVendor?: string
  rollMeter?: number
  purchaseUnitPrice?: number
  purchaseUnit?: string

  // 案件・見積
  projectNoParent?: string
  projectNoChild?: string
  projectNoMaterial?: string
  quotationNo?: string
  estimateCalcNo?: string
  refEstimateCalcNo?: string

  // セット品
  setProductCd?: string
  setProductName?: string
  parentChildDiv?: string
  setRatio?: number
  orderType?: string

  // 製品属性スナップショット
  productUsage?: string
  distributionDiv?: string
  confidentialInfo?: string
  seizureDiv?: string
  importanceDiv?: string
  mChange?: string
  qualityDiv?: string
  productShape?: string
  unescoMark?: string
  origamiMark?: string
  fourMContract?: string
  tkpWrinkleStd?: string
  recyclingPayment?: string

  paperUsageG?: number
  plasticUsageG?: number
  glassUsageG?: number
  petUsageG?: number
  packPaperUsageG?: number
  packPlasticUsageG?: number

  designProposalNo?: string
  salesPriceDiv?: string
  freightBilling?: string

  mcOrderNo?: string
  mcOrderDetailNo?: string
  status: number
  wfApprovalFlg: boolean
  mcTransferFlg: boolean

  provisionalPriceFlg: boolean
  priceChangeReason?: string
  approvalStatus?: number

  processes: OrderProcessDto[]
  processNotes: OrderProcessNoteDto[]
  materials: OrderMaterialDto[]
}

/** 受注加工工程 1 行 */
export interface OrderProcessDto {
  productCd: string
  operationCd: string
  processCd: string

  topItemCd?: string
  topBranch1?: string
  topBranch2?: string
  topBranch3?: string
  itemCd?: string
  branch1?: string
  branch2?: string
  branch3?: string

  workingGroupCd?: string
  machineOrVendor?: string
  machineFixedFlg: boolean
  cpDeliveryDiv?: string

  spec01?: string
  spec02?: string
  spec03?: string
  spec04?: string
  spec05?: string
  spec06?: string
  spec07?: string
  spec08?: string
  spec09?: string
  spec10?: string

  qtyUnit?: string
  plateNo1?: string
  plateNo2?: string
  plateNo3?: string
  consumable1?: string
  consumable2?: string
  consumable3?: string

  purchaseUnitPrice?: number
  fixedPrice?: number
  lossRate: number
  machineCount: number
  leadTimeDays: number
  storageLocation?: string

  sortOrder: number

  priorityItem1?: string
  priorityItem2?: string
  priorityItem3?: string
  priorityItem4?: string
  priorityItem5?: string
  priorityItem6?: string
  priorityItem7?: string
  priorityItem8?: string

  scheduledDate?: string
}

export interface OrderProcessNoteDto {
  productCd: string
  operationCd: string
  note1?: string
  note2?: string
}

export interface OrderMaterialDto {
  productCd: string
  processCd: string
  materialCd: string
  materialTypeDiv: string
  itemCd?: string
  branch1?: string
  branch2?: string
  branch3?: string
  supplyDiv: string
  supplyUnitPrice: number
  sortOrder: number
}

// ─────────────────── PA080 一覧 ───────────────────

export interface OrderQueryDto {
  baseCd?: string
  salesPersonCd?: string
  customerCd?: string
  customerCdTo?: string
  orderType?: string
  productCatBig?: string
  productCatMid?: string
  productCatSml?: string
  orderDateFrom?: string
  orderDateTo?: string
  deliveryDateFrom?: string
  deliveryDateTo?: string
  haibaiNo1From?: string
  haibaiNo1To?: string
  haibaiNo2From?: string
  haibaiNo2To?: string
  haibaiNo3From?: string
  haibaiNo3To?: string
  orderSheetNo?: string
  defectiveHaibaiNo?: string
  mcOrderNo?: string
  productCd?: string
  itemCd?: string
  customerItemName?: string
  customerPartNo?: string
  sheetFlute?: string
  paperCd?: string
  printCd?: string
  embossCd?: string
  makerCd?: string
  fscOrderType?: string
  salesPriceDiv?: string
  productShape?: string
  distributionDiv?: string
  carrier?: string
  onlyConsignedSales?: boolean
  onlyMcUntransferred?: boolean
  includeDeleted?: boolean
  page?: number
  pageSize?: number
  maxRows?: number
  sortField?: string
  sortOrder?: string
}

export interface OrderListItemDto {
  rowNo: number
  customerCd?: string
  salesPersonCd?: string
  customerName?: string
  salesPersonName?: string
  orderSheetNo?: string
  haibaiNo1?: string
  defectiveHaibaiNo?: string
  mcOrderNo?: string
  orderDate?: string
  customerDeliveryDate?: string
  productCd?: string
  itemCd?: string
  productCatBig?: string
  productCatBigName?: string
  cpItemOrComposition?: string
  sheetFlute?: string
  compositionF?: string
  compositionC?: string
  compositionB?: string
  compositionItemName?: string
  fullSheetWidth?: number
  fullSheetFlow?: number
  slitterWidth?: number
  slitterFlow?: number
  qtyUnit?: string
  quantity?: number
  individualUnitPrice?: number
  setUnitPrice?: number
  unitPriceUnit?: string
  amount?: number
  consignedSalesFlg?: string
  slipNote?: string
  deliveryNote?: string
  webOrderNo: string
  webOrderDetailNo: number
}

// ─────────────────── PA090 単価訂正 ───────────────────

export interface OrderPriceCorrectionQueryDto {
  baseCd?: string
  salesPersonCd?: string
  customerCd?: string
  customerCdTo?: string
  orderDateFrom?: string
  orderDateTo?: string
  deliveryDateFrom?: string
  deliveryDateTo?: string
  haibaiNo1?: string
  orderSheetNo?: string
  productCd?: string
  customerItemName?: string
  qtyFrom?: number
  qtyTo?: number
  amountFrom?: number
  amountTo?: number
  onlyProvisional?: boolean
  approvalStatus?: string
  page?: number
  pageSize?: number
  maxRows?: number
}

export interface OrderPriceCorrectionItemDto {
  rowNo: number
  approvalStatus?: number
  customerCd?: string
  customerName?: string
  salesPersonCd?: string
  salesPersonName?: string
  orderSheetNo?: string
  haibaiNo1?: string
  defectiveHaibaiNo?: string
  productCd?: string
  itemCd?: string
  cpItemName1?: string
  cpItemName2?: string
  orderDate?: string
  customerDeliveryDate?: string
  unit?: string
  cpItemOrComposition?: string
  compositionF?: string
  compositionC?: string
  compositionB?: string
  sheetFlute?: string
  fullSheetWidth?: number
  fullSheetFlow?: number
  slitterWidth?: number
  slitterFlow?: number
  quantity?: number
  qtyUnit?: string
  individualUnitPriceBefore?: number
  setUnitPriceBefore?: number
  unitPriceUnit?: string
  individualUnitPriceAfter?: number
  setUnitPriceAfter?: number
  specialPriceFlg?: string
  provisionalPriceFlg: boolean
  priceChangeReason?: string
  amount?: number
  consignedSalesFlg?: string
  slipNote?: string
  deliveryNote?: string
  webOrderNo: string
  webOrderDetailNo: number
  selected?: boolean
  rowVersion?: string
}

export interface OrderPriceCorrectionUpdateDto {
  webOrderNo: string
  webOrderDetailNo: number
  individualUnitPriceAfter?: number
  setUnitPriceAfter?: number
  specialPriceFlg?: string
  priceChangeReason?: string
  rowVersion?: string
}

export interface OrderPriceCorrectionBatchUpdateDto {
  items: OrderPriceCorrectionUpdateDto[]
}

export interface OrderPriceCorrectionBatchResultDto {
  updatedCount: number
  wfRequestedCount: number
  conflictedKeys: string[]
}

// ─────────────────── 業務ルール結果 ───────────────────

export interface IsEditableResultDto {
  isEditable: boolean
  reason?: string
}

export interface OrderWipCheckResultDto {
  level: number
  message?: string
}

export interface CreditCheckResultDto {
  isOver: boolean
  creditLimit: number
  balance: number
  message?: string
}

export interface LeadTimeRequestDto {
  shipDateTime?: string
  leadTimeDays?: number[]
  wgCd?: string
}

// ─────── Phase 6 受注取消 ───────

export type CancelOutcome = 'Cancelled' | 'NeedsDecision' | 'Rejected' | 'PartiallyCancelled'

export interface WorkOrderProbe {
  workOrderNo: string
  status: number
  autoCancellable: boolean
  cancelled?: boolean | null
  message?: string | null
}

export interface OutboundProbe {
  outboundNo: string
  status: number
  outboundType: number
  autoCancellable: boolean
  cancelled?: boolean | null
  message?: string | null
}

export interface OrderCancelResult {
  outcome: CancelOutcome
  message?: string | null
  relatedWorkOrders: WorkOrderProbe[]
  relatedOutbounds: OutboundProbe[]
  correlationId: string
}

// ─────── Phase 8 受注済未出荷 Dashboard ───────

export interface UnshippedOrderItemDto {
  webOrderNo: string
  customerCd: string
  customerName?: string | null
  orderDate?: string | null
  customerDeliveryDate?: string | null
  orderStatus: string  // CONFIRMED / IN_PRODUCTION / PARTIALLY_CANCELLED
  shipStatus: number   // 0 / 5 / 9
  orderedQty: number
  shippedQty: number
  remainingQty: number
  isOverdue: boolean
  daysUntilDue?: number | null
  mesStatusSummary?: string | null
  wmsStatusSummary?: string | null
}

export interface UnshippedOrderQuery {
  customerCd?: string
  onlyOverdue?: boolean
  page: number
  pageSize: number
  sortField?: string  // "deliveryDate" / "orderDate" / "remainingQty"
  sortOrder?: 'asc' | 'desc'
}

export interface PagedResult<T> {
  rows: T[]
  total: number
}

export type { ApiResult }
